### PR TITLE
PERF: copy cached attributes on index shallow_copy

### LINKED ISSUE #28584: PERF: Index._shallow_copy doesn't copy ._engine
```python
>>> idx = pd.Index(np.arange(100_000))
>>> %timeit idx.get_loc(99_999)
774 ns ± 26 ns per loop
>>> %timeit idx._shallow_copy().get_loc(99_999)
3.57 ms ± 56.8 µs per loop
```
The same performance issue can be seen on other index types, e.g. CategoricalIndex and MultiIndex.

#### Problem description

The reason for the above diferences is that ``_shallow_copy`` does not copy over the ``._engine`` attribute to the new index and the ``_engine`` is expensive to recreate. 

Indexes are immutable, and likewise - to my understanding - are the ``._engine`` attribute of indexes. The ``._engine`` is quite expensive to create and if it has been created on the original index, I think it should be possible to reuse it on the new index, saving ther time needed to create a new and identical ``._engine``.

``_shallow_copy`` is used in a few places internally in pandas, so there seems to be potential for some speedups for several pandas merhods by copying the _engine over to newly-copied indexes.

Possibly I'm missing some finer details here, e.g. don't know what the ``._engine.clear_mappings`` is for and it seems from its name to be destructive, but overall it seems to be possible to make a change to copy this over.

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : 79663fb669cce436c6379c83c887755780b84770
python           : 3.7.3.final.0
python-bits      : 64
OS               : Windows
OS-release       : 10
machine          : AMD64
processor        : Intel64 Family 6 Model 78 Stepping 3, GenuineIntel
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : None.None

pandas           : 0.25.0.dev0+1363.g79663fb66
numpy            : 1.16.4
pytz             : 2019.1
dateutil         : 2.8.0
pip              : 19.1.1
setuptools       : 41.0.1
Cython           : 0.29.13
pytest           : 5.0.1
hypothesis       : 4.28.2
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : None
IPython          : 7.6.1
pandas_datareader: None
bs4              : None
bottleneck       : None
fastparquet      : None
gcsfs            : None
lxml.etree       : None
matplotlib       : None
numexpr          : 2.6.9
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pytables         : None
s3fs             : None
scipy            : None
sqlalchemy       : None
tables           : None
xarray           : None
xlrd             : None
xlwt             : None
xlsxwriter       : None

</details>

### PR DESCRIPTION
**Pull request by topper-123**
- [x] closes #28584
- [ ] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

The performance of the example in #28584 is:

```python
>>> idx = pd.Index(np.arange(100_000))
>>> %timeit idx.get_loc(99_999)
1.17 µs ± 80.6 ns per loop  # master and this PR
>>> %timeit idx._shallow_copy().get_loc(99_999)
3.57 ms ± 286 ns per loop  # master
3.67 µs ± 286 ns per loop  # this PR
```

The issue is still on the extension indexes, e.g. ``CategoricalIndex._shallow_copy``. I'd like to take them afterwards.

### CODE REVIEW

File: pandas/core/indexes/base.py [hunk 1]
```diff
@@ -499,10 +501,12 @@ def _shallow_copy(self, values=None, name: Label = no_default):
         """
         name = self.name if name is no_default else name
 
-        if values is None:
-            values = self.values
+        cache = self._cache if values is None else {}
```
Comment (jbrockmendel) [inline pandas/core/indexes/base.py]: can you keep the `if values is None:` check; i like it for inspection of coverage results

**Comment by jbrockmendel** [comment]
LGTM.

**Comment by jbrockmendel** [comment]
Should engine be shared?  If so, maybe access it to ensure it is in _cache before coping _cache?

**Comment by topper-123** [comment]
If the _cache has been populated before the call to ``_shallow_copy``, we always copy. It it hasn't, it's a question of evaluating trade-offs:
1. If we access it before copying, we risk calculating too often
2. If we do not access it before copying, we risk having to do the same calculation several times.

My hunch is that it's mostly not that important, because either the cache has already been populated, or the original index will not be reused (e.g. in a pipe, often).

But in some given cases we'd prefer either 1 or 2, depending on context, but we'd have to choose. I prefer 2., because in that case a pipe will not stop while creating ``_engine`` attributes, i..e performance will be more predictable.

**Comment by jbrockmendel** [comment]
Makes sense, thanks for explaining your thought process.

**Comment by jreback** [comment]
thanks @topper-123 

if you think we need additional asvs's happy to have them (e.g. is_monotonic, is_unique), MI indexing, and .get_loc would all benefit; we might have enough, just checking in about this.

**Pull request merged**